VideoClassification/SegmentLevelClassifier/class_feature_generation.py

```diff
@@ -0,0 +1,104 @@
+"""Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    https://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Compute and add Class specific features to the data.
+"""
+import numpy as np
+import readers
+import tensorflow as tf
+import writer
+
+def calculate_cosine(segment1, segment2):
+  """Calculate the cosine of the angle between segment1 and segment2. Modified to work with matrices by applying mean of cosine similarities.
+
+  Args:
+    segment1: Matrix of vectors to compare
+    segment2: Matrix of vectors to compare
+  """
+  similarity = []
+  for i in range(len(segment1)):
+    similarity.append(np.dot(segment1[i], segment2[i]) / (np.linalg.norm(segment1[i])*np.linalg.norm(segment2[i])))
+  similarity = np.array(similarity)
+  return np.mean(similarity)
+
+def compute_and_save(data_dir, input_dir, type_comparison="train", num_classes=1000):
+  """Compute class specific features for input_dataset and save them to data_dir.
+
+  Args:
+    data_dir: directory to save data to
+    input_dir: directory where input data is stored.
+  """
+  #Store previous computations to speed up runtime
+  num_segment = 0
+  for label in range(num_classes):
+    shard = []
+    comparison_dataset_reader = readers.SegmentDataset(class_num=label)
+    comparison_dataset = comparison_dataset_reader.get_dataset("/home/conorfvedova_google_com/data/segments/split_validation", batch_size=1, type="class")
+    
+    #Preload Data and convert to numpy for calculations
+    video_holder_input = []
```

Comment:
yes it holds the examples per class. It holds the examples per class because otherwise we would perform a lot of file reads and numpy conversions within the for loop. This procedure speeds up the code by about 6x